### DISCUSSION TITLE
Bouncer TCP settings

### DISCUSSION THREAD

**Discussion by dspeck1** [cloudnative-pg/cloudnative-pg#563] [Q&A] [answered]
Trying to add these parameters to the pg bouncer config.

```
tcp_keepalive=1
tcp_keepcnt=1
tcp_keepidle=1
tcp_keepintvl=1
tcp_user_timeout=1000
```

Getting this error  `* spec.cluster.parameters: Invalid value: "tcp_keepcnt": Invalid or reserved parameter` for each value.  What is recommended way to configure these settings?  The .ini file cannot be modified directly it appears.

**Comment by leonardoce** [comment, marked as answer]
Hi!
Yes, you can't inject your init file as the operator is managing the configuration in a declarative way.

This is the list of parameters we currently allow to configure: https://github.com/cloudnative-pg/cloudnative-pg/blob/0e2bc8e8595f9443bbb25962b34d0d2530ac53ca/api/v1/pooler_webhook.go#L36

Can you please explain a bit the scenario and what you want to achieve by changing those parameters?

**Reply by yee379** [reply]
we're hitting timeouts as documented at https://tech.instacart.com/the-vanishing-thread-and-postgresql-tcp-connection-parameters-93afc0e1208c and wish to implement these on the server/pooler side.

**Reply by leonardoce** [reply]
That's a 502 for me :(

**Reply by yee379** [reply]
indeed... :( although it seems a bit more stable now...

**Reply by leonardoce** [reply]
I see now.... anyway I'm not so sure in Kubernetes they'll have the same meaning.
We are working with K8s services here, and kube-proxy will certainly play a role. Look for the TCP network parameters in the [kube-proxy documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy).
What do you think about it?

**Reply by yee379** [reply]
thanks @leonardoce - interesting idea to get it into kube-proxy rather than in the pooler. however, i do not see an exact mapping between the parameters above and that in kube-proxy. i will note that we did apply the above parameters to a zalando pgbouncer setup (on the same k8s cluster) and the client timeouts went away - so it would be great to be able to apply these declaratively in the cnpg pooler.

**Reply by leonardoce** [reply]
Yes I agree with you, there's no direct relationship between them and the kube-proxy ones are definitely more broadly applied.
Anyway, I'm unfortunately not so familiar with the pooler implementation of the Zalando operator. Does the Zalando operator rely on a service between the Pooler and PostgreSQL to forward connections?

@yee379 would you like to contribute a patch to allow the users to customize them? At first glance adding those parameters in the ones we allow to change should be enough.

You'll need to create an issue referencing this discussion and then a corresponding PR. More info here: https://github.com/cloudnative-pg/cloudnative-pg/blob/main/CONTRIBUTING.md

After that I'll be more than happy to review the PR for you. Thanks!

**Reply by dspeck1** [reply]
Thanks @leonardoce   here is PR.  https://github.com/cloudnative-pg/cloudnative-pg/pull/584

**Reply by leonardoce** [reply]
Great! Thank you!
